### DISCUSSION TITLE
More details on Microcks own's API

### DISCUSSION THREAD

**Discussion by asankas** [microcks/microcks#1155] [Community]
Hi, 
I would like to know whether there are APIs (https://microcks.io/documentation/automating/api/) exposed by microcks which can be used in programatically upload api specification to generate mocks. I was trying some APIs mentioned in above link and ended up with 401 response. I would like to know what kind of authorisation should be provided when accessing those APIs (Note: I was trying with Basic Auth with both 'microcks-app' and 'microcks-serviceaccount' client IDs with corresponding client secrets from keycloak). Appreciate if anyone from the community can provide more details on how to use those microcks own APIs. 

Regards,
Asanka.

**Comment by lbroudoux** [comment]
Hey @asankas! 

Some endpoints exist at the moment but are not documented as they were not that stable. Now, as they are used more and more by CLI, Testcontainers module or CI plugins, they'll be documented in future releases.
You can currently use:
* `POST /api/artifacts/upload?mainArtifact=true|false` with a `multipart/form-data` payload wrapping your artifact file OR
* `POST /api/artifacts/download?mainArtifact=true|false&url=<artifact_url>` to force Microcks to import a remote artifact.

Or you can use the CLI that does the same thing for you ;-) 

Regarding API authentication, please check https://microcks.io/documentation/automating/service-account/#using-service-account to see how to use the service-account client ID and secrets to retrieve a token that will then allow you to interact with the API. Take care: the default service account doesn't allow you to import artifacts (as mentioned [here](https://microcks.io/documentation/automating/service-account/#inspecting-default-service-account)). You'll have to extend its privileges or create a new one.

**Reply by asankas** [reply]
Hi Ibroudoux,

Thank you for the prompt response. 

We are trying out microcks in one of our project called [xapi hub](https://xapihub.io/), which provides tooling platform the compose api specification (something similar to swaggerhub). We are doing a POC whether we can use microcks to generate mocks for the api specifications composed through our tooling platform. Basically what we need is some APIs to upload the specification to microcks from our tool via an API. Then it should return some endpoints to tryout those mocks. 

We expect to disable default authorisation/authentication comes in microcks (via keycloak) as we called microcks APIs from our secured backend. 

In your previous reply you mentioned that _"the default service account doesn't allow you to import artifacts. You'll have to extend its privileges or create a new one"_, Can you please explain what privileges it requires and how it can be extended. Appreciate if you can point a detail documentation for this. 

We would like to know whether microcks has the flexibility for our requirement. Kindly advise. 

Thank You and Regards,
Asanka.

**Reply by lbroudoux** [reply]
Your project looks really nice!

I think you'd like to check https://microcks.io/documentation/automating/service-account/#inspecting-default-service-account to see how default service account is defined. The required role for using the `/api/artifacts/upload` and `/api/artifacts/download` is the `manager` role that is defined on the `microcks-app` resource within the Keycloak realm.
